src/js/media-controller.js

```diff
@@ -65,60 +70,96 @@ class MediaController extends MediaContainer {
-        e.stopPropagation();
+    let prevState = {};
+    this.#mediaStateCallback = (nextState) => {
+      Object.entries(nextState).forEach(([stateName, stateValue]) => {
+        // Make sure to propagate initial state, even if still undefined (CJP)
+        if (stateName in prevState && prevState[stateName] === stateValue) return;
+        this.propagateMediaState(stateName, stateValue);
+        const attrName = stateName.toLowerCase();
+        // TODO: I don't think we want these events to bubble? Video element states don't. (heff)
+        const evt = new globalThis.CustomEvent(
+          AttributeToStateChangeEventMap[attrName],
+          { composed: true, bubbles: true, detail: stateValue }
+        );
 
-        if (!this.media) {
-          console.warn('MediaController: No media available.');
-          return;
-        }
+        this.dispatchEvent(evt);
+      });
+      prevState = nextState;
+    };
 
-        MediaUIRequestHandlers[key](this.media, e, this);
-      };
-      this.addEventListener(MediaUIEvents[key], this[handlerName]);
-    });
+    /** @TODO rootNode was refactored to get set/updated in connectedCallback. Figure out what needs to change for store setup for parity */
+    if (!this.hasAttribute(Attributes.NO_DEFAULT_STORE)) {
+      this.#setupDefaultStore();
+    }
 
-    // Build event listeners for media states
-    this._mediaStatePropagators = {};
-    Object.keys(MediaUIStates).forEach((key) => {
-      this._mediaStatePropagators[key] = e => {
-        this.propagateMediaState(MediaUIProps[key], MediaUIStates[key].get(this, e));
-      };
+    this.enableHotkeys();
+  }
+
+  #setupDefaultStore() {
+    this.mediaStore = createMediaStore({
+      media: this.media,
+      rootNode: this.#rootNode,
+      fullscreenElement: this.fullscreenElement,
+      options: {
+        defaultSubtitles: this.hasAttribute(Attributes.DEFAULT_SUBTITLES),
+        defaultDuration: this.hasAttribute(Attributes.DEFAULT_DURATION) ? +this.getAttribute(Attributes.DEFAULT_DURATION) : undefined,
+        defaultStreamType: /** @type {import('./mediaStore.js').StreamTypeValue} */ (this.getAttribute(Attributes.DEFAULT_STREAM_TYPE)) ?? undefined,
+        liveEdgeOffset: this.hasAttribute(Attributes.LIVE_EDGE_OFFSET) ? +this.getAttribute(Attributes.LIVE_EDGE_OFFSET) : undefined,
+        // NOTE: This wasn't updated if it was changed later. Should it be? (CJP)
+        noVolumePref: this.hasAttribute(Attributes.NO_VOLUME_PREF),
+        noSubtitlesLangPref: this.hasAttribute(Attributes.NO_SUBTITLES_LANG_PREF),
+      },
+      // NOTE: When using a default store, we're going to assume that we should stop monitoring the state owners whenever
+      // there are no subscribers. This allows us to do proper teardown/GC when the media-controller is no longer in use.
+      /** @TODO Re-fix me! (current solution may be buggy) (CJP) */
+      monitorStateOwnersOnlyWithSubscriptions: true,
```

Comment:
should this looked at before merging?

Reply:
Should be fixed but yes I should revisit 💯